docs/design/loadpath.md

```diff
@@ -0,0 +1,284 @@
+# bpfman outside of the load-path
+
+## Introduction
+
+One of the drawbacks of bpfman in it's current incarnation is that it requires
+users to "opt in" to our opinionated loading pattern for eBPF programs.
+
+From a "linux only" standpoint the benefits of our scheme are as follows:
+
+1. Offering containers as a lightweight packaging alternative to RPMs/DEBs
+   for eBPF programs.
+1. Provides signing of eBPF Container Images through cosign.
+1. Provides a way to load eBPF programs from a container image without
+   requiring the container runtime to be installed on the host.
+1. Allows for long-running user-space processes (that work with eBPF maps) to
+   be run with no privileges, since most map read interactions are permissable
+   without CAP_BPF. This is a security benefit, since it reduces the attack
+   surface of the user-space process. There are a few exceptions to this
+   rule (e.g perf_event_array maps), but we're planning on finding a way to
+   work around this.
+1. Simplified (and consistent) lifecycle management of eBPF programs regardless
+   of the program type or attachment point.
+
+Given that a lot of the software using eBPF is also Cloud-Native, we have the
+following additional benefits:
+
+1. A Cloud-Native, declarative way to load eBPF programs and have them
+   orchestrated by Kubernetes. This is backed by packaging eBPF programs as
+   container images.
+1. For daemonsets/deployments that use these eBPF programs to not require
+   privileged containers, host mounts or any capabilities.
+
+These benefits however are proving to be insufficient to sway prospective users
+away from loading their own eBPF programs for 2 reasons:
+
+1. They are an established project and has code that works and don't see a
+   reason to change it
+1. They believe that separating the user-space code and kernel-space code is
+   a bad idea, citing the following reasons:
+   - The versions of the kernel and user-space must be tightly coupled.
+   - They are wedded to BPF Skeletons (bpftool generated C code) since they
+     require complex .bss/.rodata section manipulation before loading, and they
+     argue that this is too complex to model via an API.
```

Comment:
There_ are applications of dynamically generating BPF bytecode based on user input: https://github.com/xdp-project/xdp-tools/pull/356 - so I believe "dynamically generated code" should go on this list as well (BCC comes to mind as another user of this)...